### PR TITLE
Add Hugging Face Hub LLM support

### LINKED ISSUE #686: Support HF Hub and HF local pipeline via langchain
### 🚀 The feature

https://python.langchain.com/docs/integrations/llms/huggingface_hub
https://python.langchain.com/docs/integrations/llms/huggingface_pipelines

### Motivation, pitch

* We need to support more llms so that users have the flexibility of choosing whatever LLM they want to use.
* Langchain provides a standard interface for llms. This helps in better code maintenance and also helps in better observability.

### PR DESCRIPTION
**Pull request by sidmohanty11**
## Description

This PR adds Hugging Face Hub LLM support to Embedchain

Fixes #686

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (does not change functionality, e.g. code style improvements, linting)
- [ ] Documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

Please delete options that are not relevant.

- [x] Unit Test
- [x] Test Script (please provide)

```python
os.environ["OPENAI_API_KEY"] ="sk-xxx"
os.environ["HUGGINGFACEHUB_API_TOKEN"] = "hf_xxx"

app = App(
    llm=HuggingFaceHubLlm(),
)

app.add("https://en.wikipedia.org/wiki/Pok%C3%A9mon")

config = BaseLlmConfig(top_p=0.1,model="mistralai/Mistral-7B-v0.1", max_tokens=100)

response = app.query("What is a pokemon?", config=config)

print("Answer: ", response)

```

```
Output:
Answer:  
  A pokemon is a creature that can be caught and trained to fight in battles.

  Helpful Answer:

  A pokemon is a creature that can be caught and trained to fight in battles.

  Helpful Answer:

  A pokemon is a creature that can be caught and trained to fight in battles.

  Helpful Answer:

  A pokemon is a creature that can be caught and trained to fight in battles.
```

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

## Maintainer Checklist

- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] Made sure Checks passed

### CODE REVIEW

File: embedchain/llm/hugging_face_hub.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+from embedchain.config import BaseLlmConfig
+from embedchain.helper.json_serializable import register_deserializable
+from embedchain.llm.base import BaseLlm
+
+
+@register_deserializable
+class HuggingFaceHubLlm(BaseLlm):
+    def __init__(self, config: Optional[BaseLlmConfig] = None):
+        if "HUGGINGFACEHUB_API_TOKEN" not in os.environ:
```
Comment (deshraj) [inline embedchain/llm/hugging_face_hub.py:15]:
```suggestion
        if "HUGGINGFACEHUB_ACCESS_TOKEN" not in os.environ:
```

File: embedchain/llm/hugging_face_hub.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+            raise ValueError("Please set the HUGGINGFACEHUB_API_TOKEN environment variable.")
```
Comment (deshraj) [inline embedchain/llm/hugging_face_hub.py:16]:
```suggestion
            raise ValueError("Please set the HUGGINGFACEHUB_ACCESS_TOKEN environment variable.")
```

File: embedchain/llm/hugging_face_hub.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+            "max_new_tokens": config.max_tokens,
+        }
+
+        if config.top_p > 0.0 and config.top_p < 1.0:
+            model_kwargs["top_p"] = config.top_p
+        else:
+            raise ValueError("`top_p` must be > 0.0 and < 1.0")
+
+        llm = HuggingFaceHub(
+            huggingfacehub_api_token=os.environ["HUGGINGFACEHUB_API_TOKEN"],
```
Comment (deshraj) [inline embedchain/llm/hugging_face_hub.py:46]:
```suggestion
            huggingfacehub_api_token=os.environ["HUGGINGFACEHUB_ACCESS_TOKEN"],
```

File: tests/llm/test_hugging_face_hub.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+import unittest
+from unittest.mock import patch
+
+from embedchain.config import BaseLlmConfig
+from embedchain.llm.hugging_face_hub import HuggingFaceHubLlm
+
+
+class TestHuggingFaceHubLlm(unittest.TestCase):
+    def setUp(self):
+        os.environ["HUGGINGFACEHUB_API_TOKEN"] = "test_api_key"
```
Comment (deshraj) [inline tests/llm/test_hugging_face_hub.py:12]:
```suggestion
        os.environ["HUGGINGFACEHUB_ACCESS_TOKEN"] = "test_access_token"
```

File: tests/llm/test_hugging_face_hub.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        self.config = BaseLlmConfig(model="google/flan-t5-xxl", max_tokens=50, temperature=0.7, top_p=0.8)
+
+    def test_init_raises_value_error_without_api_key(self):
+        os.environ.pop("HUGGINGFACEHUB_API_TOKEN")
```
Comment (deshraj) [inline tests/llm/test_hugging_face_hub.py:16]:
```suggestion
        os.environ.pop("HUGGINGFACEHUB_ACCESS_TOKEN")
```

File: tests/llm/test_hugging_face_hub.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+        with self.assertRaises(ValueError):
+            HuggingFaceHubLlm()
+
+    def test_get_llm_model_answer_raises_value_error_for_system_prompt(self):
+        llm = HuggingFaceHubLlm(self.config)
```
Comment (deshraj) [inline tests/llm/test_hugging_face_hub.py:21]: I believe these tests will only pass if we supply the `HUGGINGFACEHUB_ACCESS_TOKEN` env variable. Can you please mock the  actual hugging face object that you are using in `HuggingFaceHubLlm` code?

**Review by deshraj — COMMENTED** [review]
This is great! Please address the minor comments I have and we are good to go.

File: tests/llm/test_hugging_face_hub.py (hunk 6 above)
Comment (sidmohanty11) [reply tests/llm/test_hugging_face_hub.py:21]: Thanks a lot for the feedback @deshraj! I've added a test mocking the actual HuggingFace object. Please let me know if this aligns with the intended result.

**Review by deshraj — APPROVED** [review]

**Pull request merged**